elastic/beats
# Fix issues with output reloading

## What does this PR do?

Beats Central Management (today) and Agent (in the future) will need to reload outputs on the fly, whenever users change the Beat's configuration while the Beat is running. The current output reloading implementation has three bugs:

1. When outputs are reloaded several times, and very quickly at that, all events don't always getting eventually published,

1. Whenever an output is reloaded, a goroutine is leaked from the publisher pipeline's output handling code, and

2. If the output being reloaded is the Elasticsearch output, old TCP connections to Elasticsearch are hanging around.

This PR addresses all three issues.

## Why is it important?

- To ensure that all events will eventually be published, even if outputs are rapidly reloaded multiple times.
- To prevent leakage of memory (goroutines consume memory).
- To prevent exhaustion of file descriptors over time.

## Checklist


## How to test this PR locally

1. Start up Elasticsearch and Kibana and setup [Beats Central Management](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-central-management.html).
   1. Enroll Filebeat. This will backup your original `filebeat.yml` file and create a new one suitable for Beats Central Management.
   1. Create a configuration tag in the Beats Central Management UI in Kibana with input paths containing `/tmp/logs/*.log` and the Elasticsearch output pointing to localhost:9200 (or wherever your local ES is listening). Make sure to assign this tag to your enrolled Filebeat instance in the UI.
3. Edit the new `filebeat.yml` file and set `management.period: 2s` to speed up output reloading.
4. Start up Filebeat with the `-httpprof localhost:5050` flag. This will let us see if Filebeat is leaking goroutines.
5. In a new window, check TCP connections used by your Filebeat process. This will let us see if Filebeat is leaking file descriptors.
   ```
   watch -n 2 lsof -a -cfilebeat -i4 -i6 -itcp
   ```
6. In another new window, make sure some data is constantly being ingested by Filebeat from the input path.
   ```
   rm -rf /tmp/logs
   mkdir -p /tmp/logs
   while true; do echo "foo bar $RANDOM" >> /tmp/logs/test.log; sleep 1; done
   ```
7. In the Central Management UI, change the configuration for your tag slightly (maybe add an input path then later remove it and so on) to trigger output reloading in Filebeat. 
8. Check the Filebeat log for messages like `New configurations retrieved` to confirm that your changes in Kibana are being picked up by Filebeat.
8. Check the output of the `lsof` window and make sure the list of file descriptors isn't growing over time.
9. Visit (or refresh) http://localhost:5050/debug/pprof/goroutine?debug=1 in a browser window and make sure the number of goroutines isn't growing over time.
10. Wait a few seconds, then repeat steps 6 onwards a few times.

## Related issues

- Fixes elastic/beats#10491
